app/models/concerns/browseable.rb

```diff
@@ -0,0 +1,157 @@
+module Browseable
+  extend ActiveSupport::Concern
+
+  included do
+    class_attribute :facets
+    self.facets = {}
+  end
+
+  class Facets
+    include Enumerable
+
+    attr_reader :search, :klass
+
+    def initialize(search)
+      @search, @klass = search, search.klass
+    end
+
+    def [](key)
+      facets[key]
+    end
+
+    def key?(key)
+      facet?(key)
+    end
+    alias_method :has_key?, :key?
+
+    def keys
+      klass.facets.keys
+    end
+
+    def each(&block)
+      keys.each do |key|
+        yield key, self[key]
+      end
+    end
+
+    private
+
+    def facet(key)
+      klass.facets.fetch(key)
+    end
+
+    def facet?(key)
+      klass.facets.key?(key)
+    end
+
+    def facets
+      @facets ||= Hash.new(&facet_query)
+    end
+
+    def facet_query
+      lambda do |hash, key|
+        unless facet?(key)
+          raise ArgumentError, "Unsupported facet: #{key.inspect}"
+        end
+
+        hash[key] = scope(key).count
+      end
+    end
+
+    def scope(key)
+      search.klass.instance_exec(&facet(key))
+    end
+  end
```

Comment:
:+1: there is a small performance penalty for using delegate but not worth worrying about